Sources/swift-format/Frontend/Frontend.swift

```diff
@@ -124,27 +124,42 @@ class Frontend {
   }
 
   /// Processes source content from a list of files and/or directories provided as paths.
-  private func processPaths(_ paths: [String]) {
+  private func processPaths(_ paths: [String], parallel: Bool) {
     precondition(
       !paths.isEmpty,
       "processPaths(_:) should only be called when paths is non-empty.")
 
-    for path in FileIterator(paths: paths) {
+    let openAndProcess: (String, NSLock?) -> Void = { path, lock in
       guard let sourceFile = FileHandle(forReadingAtPath: path) else {
-        diagnosticEngine.diagnose(Diagnostic.Message(.error, "Unable to open \(path)"))
-        continue
+        lock?.lock()
+        defer { lock?.unlock() }
+        self.diagnosticEngine.diagnose(Diagnostic.Message(.error, "Unable to open \(path)"))
+        return
       }
 
-      guard let configuration = configuration(
-        atPath: lintFormatOptions.configurationPath, orInferredFromSwiftFileAtPath: path)
+      guard let configuration = self.configuration(
+        atPath: self.lintFormatOptions.configurationPath, orInferredFromSwiftFileAtPath: path)
       else {
         // Already diagnosed in the called method.
-        continue
+        return
       }
 
       let fileToProcess = FileToProcess(
         fileHandle: sourceFile, path: path, configuration: configuration)
-      processFile(fileToProcess)
+      self.processFile(fileToProcess)
+    }
+
+    let lock = NSLock()
+    let allFilePaths = Array(FileIterator(paths: paths))
```

Comment:
Should we go back to the other option then? Not sure I see a big win from using `concurrentPerform` at the moment especially if it requires this extra branching

Reply:
(Pushed this change in the meantime)

Reply:
That's why I was curious about `OperationQueue`, because it seems like it still provides a way to let the system choose the parallelism while also letting us walk the file hierarchy lazily, by just calling `addOperation` repeatedly instead of requiring the iteration count up front. 🤷‍♂ 

But if that's not advisable for some reason, I don't consider pre-computing the file list only in parallel mode to be extra branching—rather, it's choosing the best distribution of the work given the requirements of the APIs we're using (or not using).